src/Internal/Contract/Monad.purs

```diff
@@ -0,0 +1,400 @@
+derive newtype instance MonadAff Contract
+derive newtype instance Semigroup a => Semigroup (Contract a)
+derive newtype instance Monoid a => Monoid (Contract a)
+derive newtype instance MonadRec Contract
+derive newtype instance MonadAsk ContractEnv Contract
+derive newtype instance MonadReader ContractEnv Contract
+-- Utilise JavaScript's native `Error` via underlying `Aff` for flexibility:
+derive newtype instance MonadThrow Error Contract
+derive newtype instance MonadError Error Contract
+
+instance MonadLogger Contract where
+  log msg = do
+    config <- ask
+    let logFunction = fromMaybe logWithLevel config.customLogger
+    liftAff $ logFunction config.logLevel msg
+
+instance Parallel ParContract Contract where
+  parallel :: Contract ~> ParContract
+  parallel (Contract a) = ParContract $ parallel a
+  sequential :: ParContract ~> Contract
+  sequential (ParContract a) = Contract $ sequential a
+
+newtype ParContract (a :: Type) = ParContract
+  (ReaderT ContractEnv ParAff a)
+
+derive newtype instance Functor ParContract
+derive newtype instance Apply ParContract
+derive newtype instance Applicative ParContract
+derive newtype instance Alt ParContract
+derive newtype instance Plus ParContract
+derive newtype instance Alternative ParContract
+derive newtype instance Semigroup a => Semigroup (ParContract a)
+derive newtype instance Monoid a => Monoid (ParContract a)
+
+-- | Interprets a contract into an `Aff` context.
+-- | Implicitly initializes and finalizes a new `ContractEnv` runtime.
+-- |
+-- | Use `withContractEnv` if your application contains multiple contracts that
+-- | can be run in parallel, reusing the same environment (see
+-- | `withContractEnv`)
+runContract :: forall (a :: Type). ContractParams -> Contract a -> Aff a
+runContract params contract = do
+  withContractEnv params \config ->
+    runContractInEnv config contract
+
+-- | Runs a contract in existing environment. Does not destroy the environment
+-- | when contract execution ends.
+runContractInEnv :: forall (a :: Type). ContractEnv -> Contract a -> Aff a
+runContractInEnv contractEnv =
+  flip runReaderT contractEnv <<< unwrap
+
+--------------------------------------------------------------------------------
+-- ContractEnv
+--------------------------------------------------------------------------------
+
+type ContractEnv =
+  { backend :: QueryBackends QueryBackend
+  -- ctlServer is currently used for applyArgs, which is needed for all backends. This will be removed later
+  , ctlServerConfig :: Maybe ServerConfig
+  , networkId :: NetworkId
+  , logLevel :: LogLevel
+  , walletSpec :: Maybe WalletSpec
+  , customLogger :: Maybe (LogLevel -> Message -> Aff Unit)
+  , suppressLogs :: Boolean
+  , hooks :: Hooks
+  , wallet :: Maybe Wallet
+  , usedTxOuts :: UsedTxOuts
+  -- ledgerConstants are values that technically may change, but we assume to be
+  -- constant during Contract evaluation
+  , ledgerConstants ::
```

Comment:
In principle, we shouldn't have problems with this unless someone tries to use a test blockchain with dynamic changes in protocol parameters. So we should at least add a little warning in the documentation of some functions like `waitUntilSlot`

Reply:
I mean that the most realistic case when this can be a problem for the users is when they are using a custom network and doing something like updating the network after the initialization of the Contract env. I don't think this would be a major problem but maybe we need to document that it could happen.

Reply:
I think adding comments to functions which depend on a protocol parameters would be messy, the balancer is another example, there's quite a few more. I'm not sure how other tools deal with it, but I think we could at least add it to the general docs/faq of CTL. Note that this behaviour is already what happens in `develop`, so documentation about this doesn't depend on this PR

Reply:
This is very much an edge case scenario and so I agree that commenting the actual code is not needed.